almost_tcp/message_hdl.py

```diff
@@ -0,0 +1,249 @@
+
+Almost TCP is a protocol designed to tunnel data streams
+(specifically, HTTP requests) over a serial port.
+It consists of a subset of TCP's functionality,
+including reduced sizes for some fields,
+with the goal of enabling a small hardware implementation.
+
+Each ATCP message consists of a fixed-length header
+followed by variable-length data.
+
+All numeric fields are network-endian (big-endian).
+
+-   Connection-oriented: syn/synack/ack setup
+-   Single stream number, not two port numbers
+    -   Stream number probing to determine # of hardware resources available
+
+"""
+
+from amaranth import Module, Signal, unsigned, Const
+from amaranth.lib.wiring import Component, In, Out, Signature
+from amaranth.lib import stream
+from amaranth.lib.data import UnionLayout, ArrayLayout, Struct
+from amaranth.lib.fifo import SyncFIFOBuffered
+
+
+class FlagsLayout(Struct):
+    """
+    Layout of the flags in an ATCP header.
+    """
+
+    # Indicates the sender has sent all the data it will send on this stream.
+    fin: 1
+    # Indicates a desire to open a new stream,
+    # starting at 1 more than the provided sequence number.
+    syn: 1
+    # Indicates the provided stream should be immediately reset.
+    rst: 1
+    # Unused: push data to the receiver.
+    psh: 1
+    # The acknowledgement number is significant.
+    ack: 1
+    # Unused: urgent
+    urg: 1
+    # Unused: explicit congestion notification
+    ecn: 1
+    # Unused: congestion window reduced
+    cwr: 1
+
+
+class HeaderLayout(Struct):
+    """
+    Layout of a full ATCP header.
+
+    TODO: Fix byte order!
+    """
+
+    flags: FlagsLayout
+    # Stream: identity of the stream this message is intended for.
+    stream: 8
+
+    # Length of the data that follows the header.
+    length: 16
+
+    # Amount of additional data (after the acknowledged data)
+    # the sender can accept before buffering.
+    window: 16
+
+    # Sequence number: number in this stream of the
+    # first octet in the data section.
+    # The "syn" message is considered to occupy the first octet of the stream.
+    seq: 16
+    # Acknowledgement number: the next number the sender of this ACK expects.
+    ack: 16
+
+
+class PacketSignature(Signature):
+    """
+    Signature of a packet producer.
+
+    Attributes
+    ----------
+    header:         Header, out
+                    The header curently read in to the buffer.
+    stream_valid:   Out(1)
+                    High if the "stream" value from the header is valid.
+    header_valid:   Out(1)
+                    High if the whole header is valid.
+
+    data:           Stream(8), out
+                    Output stream for the body of the packet.
+                    The valid bits automatically deassert after the body
+                    has been fully read from the data stream.
+    """
+
+    def __init__(self):
+        super().__init__({
+            "input": In(stream.Signature(8)),
```

Comment:
Missing from attributes? Doesn't look to be used in ReadPacketStop. Is this something left over, or something you plan on using?

Reply:
Removed. Yeah, this was left over: I had copied these from an earlier iteration of `ReadPacketStop` that had implementation & interface.